File: backend/domain/workflow/entity/message.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"github.com/cloudwego/eino/schema"

	"github.com/coze-dev/coze-studio/backend/api/model/workflow"
	"github.com/coze-dev/coze-studio/backend/domain/workflow/entity/vo"
)

type Message struct {
	*StateMessage
	*DataMessage
}

// StateMessage represents a status change for the workflow execution.
type StateMessage struct {
	ExecuteID      int64
	EventID        int64 // the resuming event ID for current execution
	SpaceID        int64
	Status         WorkflowExecuteStatus
	Usage          *TokenUsage
	LastError      vo.WorkflowError
	InterruptEvent *InterruptEvent
}

// DataMessage represents a full or chunked message during a run that should go into message history.
type DataMessage struct {
	ExecuteID    int64 // the root execute ID for current execution
	Role         schema.RoleType
	Type         MessageType
	Content      string
	NodeID       string
	NodeTitle    string
	NodeType     NodeType
	Last         bool
	Usage        *TokenUsage
	FunctionCall *FunctionCallInfo
	ToolResponse *ToolResponseInfo
}

type MessageType string

const (
	Answer       MessageType = "answer"
	FunctionCall MessageType = "function_call"
	ToolResponse MessageType = "tool_response"
)

type FunctionInfo struct {
	Name string   `json:"name"`
	Type ToolType `json:"plugin_type"`

	PluginID   int64  `json:"plugin_id,omitempty"`
	PluginName string `json:"plugin_name,omitempty"`
	APIID      int64  `json:"api_id,omitempty"`
	APIName    string `json:"api_name,omitempty"`

	WorkflowName          string           `json:"workflow_name,omitempty"`
	WorkflowTerminatePlan vo.TerminatePlan `json:"terminate_plan,omitempty"`
}

type FunctionCallInfo struct {
	FunctionInfo
	CallID    string         `json:"-"`
	Arguments map[string]any `json:"arguments"`
}

type ToolResponseInfo struct {
	FunctionInfo
	CallID   string
	Response string
}

type ToolType = workflow.PluginType

const (
	PluginTool   = workflow.PluginType_PLUGIN
	WorkflowTool = workflow.PluginType_WORKFLOW
)

File: frontend/packages/studio/open-platform/open-chat/src/chat/builder-chat/plugins/event-callback/services/life-cycle/message.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type WriteableMessageLifeCycleServiceGenerator } from '@coze-common/chat-area';

import { type PluginBizContext } from '../../types/biz-context';

export const messageLifeCycleServiceGenerator: WriteableMessageLifeCycleServiceGenerator<
  PluginBizContext
> = plugin => {
  let hasInit = false;
  let lastMessageId = '';
  let nowReceivingReplyId = '';
  return {
    onBeforeMessageGroupListUpdate: ctx => {
      const { messages } =
        plugin.chatAreaPluginContext.readonlyAPI.message.getMessagesStoreInstantValues();
      const latestMessage = messages?.[0];
      if (!hasInit || lastMessageId === latestMessage?.message_id) {
        hasInit = true;
        return ctx;
      }
      plugin.pluginBizContext.eventCallbacks?.onMessageChanged?.();
      lastMessageId = latestMessage?.message_id;

      if (
        (latestMessage?.type === 'answer' &&
          latestMessage?.is_finish === true) ||
        !lastMessageId
      ) {
        plugin.pluginBizContext?.eventCallbacks?.onMessageReceivedFinish?.();
      }
      return ctx;
    },
    onAfterSendMessage: ctx => {
      const chatflowExecuteId: string =
        // @ts-expect-error -- linter-disable-autofix, 新添加参数，接口未支持
        ctx?.message?.extra_info?.chatflow_execute_id || '';
      if (chatflowExecuteId) {
        plugin.pluginBizContext?.eventCallbacks?.onGetChatFlowExecuteId?.(
          chatflowExecuteId,
        );
      }
      plugin.pluginBizContext?.eventCallbacks?.onMessageSended?.();
    },
    onBeforeReceiveMessage: ctx => {
      if (nowReceivingReplyId === ctx.message.reply_id) {
        return;
      }
      nowReceivingReplyId = ctx.message.reply_id;
      plugin.pluginBizContext?.eventCallbacks?.onMessageReceivedStart?.();
    },
  };
};
